src/coreclr/jit/lower.cpp

```diff
@@ -8179,8 +8185,354 @@ void Lowering::LowerIndir(GenTreeIndir* ind)
+                    // *(x+4), *(x+8), *(x+0), *(x+12)
+                    prev.Indir = nullptr;
+                    result     = true;
+                    break;
+                }
+                break;
+            }
+            else
+            {
+                JITDUMP("  ..but at wrong offset\n");
+            }
+        }
+    }
+
+    m_blockIndirs.Push(newInd);
+    return result;
+}
+
+//------------------------------------------------------------------------
+// TryMakeIndirsAdjacent: Try to prove that it is legal to move an indirection
+// to be adjacent to a previous indirection. If successful, perform the move.
+//
+// Arguments:
+//    ind   - Indirection to record and to try to move.
+//    next  - [out] Next node to lower. Only written if the optimization was successful.
+//
+// Returns:
+//    True if the optimization was successful.
+//
+bool Lowering::TryMakeIndirsAdjacent(GenTreeIndir* prevIndir, GenTreeIndir* indir)
+{
+    if ((indir->gtFlags & GTF_IND_VOLATILE) != 0)
+    {
+        JITDUMP("  ..but second indir is volatile\n");
+        return false;
+    }
+
+    GenTree* cur = prevIndir;
+    for (int i = 0; i < 16; i++)
+    {
+        cur = cur->gtNext;
+        if (cur == indir)
+            break;
+    }
+
+    if (cur != indir)
+    {
+        JITDUMP("  ..but it is too far away\n");
+        return false;
+    }
+
+    JITDUMP(
+        "  ..and it is close by. Trying to move the following range (where * are nodes part of the data flow):\n\n");
+#ifdef DEBUG
+    bool     isClosed;
+    GenTree* startDumpNode = BlockRange().GetTreeRange(prevIndir, &isClosed).FirstNode();
+    GenTree* endDumpNode   = indir->gtNext;
+
+    auto dumpWithMarks = [=]() {
+        if (!comp->verbose)
+        {
+            return;
+        }
+
+        for (GenTree* node = startDumpNode; node != endDumpNode; node = node->gtNext)
+        {
+            const char* prefix;
+            if (node == prevIndir)
+                prefix = "1. ";
+            else if (node == indir)
+                prefix = "2. ";
+            else if ((node->gtLIRFlags & LIR::Flags::Mark) != 0)
+                prefix = "*  ";
+            else
+                prefix = "   ";
+
+            comp->gtDispLIRNode(node, prefix);
+        }
+    };
+
+#endif
+
+    MarkTree(indir);
+
+    INDEBUG(dumpWithMarks());
+    JITDUMP("\n");
+
+    m_scratchSideEffects.Clear();
+
+    for (GenTree* cur = prevIndir->gtNext; cur != indir; cur = cur->gtNext)
+    {
+        if (cur->IsCall() || (cur->OperIsStoreBlk() && (cur->AsBlk()->gtBlkOpKind == GenTreeBlk::BlkOpKindHelper)))
+        {
+            JITDUMP("[Heuristic] Giving up due to node that kills registers [%06u]\n", Compiler::dspTreeID(cur));
+            UnmarkTree(indir);
+            return false;
+        }
+
+        if ((cur->gtLIRFlags & LIR::Flags::Mark) != 0)
+        {
+            // Part of data flow of 'indir', so we will be moving past this node.
```

Comment:
```suggestion
            // Part of data flow of 'indir', so we will be moving it upward past this node.
```

Reply:
The suggestion doesn't seem quite right to me, but I rephrased it in a different way which hopefully makes more sense.